alt-tab-macos/ui/ThumbnailsPanel.swift

```diff
@@ -91,8 +91,9 @@ class ThumbnailsPanel: NSPanel, NSCollectionViewDataSource, NSCollectionViewDele
         }
     }
 
-    func computeThumbnails() {
-        let maxSize = NSSize(width: NSScreen.main!.frame.width * Preferences.maxScreenUsage!, height: NSScreen.main!.frame.height * Preferences.maxScreenUsage!)
+    func computeThumbnails(_ screenOverride: NSScreen? = nil) {
+        let preferredScreen = screenOverride ?? Screen.getPreferredScreen() // use optional provided screen or determine it just in time
+        let maxSize = NSSize(width: preferredScreen.frame.width * Preferences.maxScreenUsage!, height: preferredScreen.frame.height * Preferences.maxScreenUsage!)
```

Comment:
I think the `maxSize` for the main panel, as well as the main panel origin code inside `showCenteredFrontPanel` would be good candidates to be extracted in the `Screen` class. I'm trying to keep the `Application` and `ThumbnailsPanel` from becoming god classes since most everything happens there. As an anecdote, I started with only `Application` and it quickly became clear that I had to split stuff :p

Reply:
**Moved to Screen**
- maxSize -> calcFrameMaxSize()
- showCenteredFrontPanel

Reply:
> Btw. does the maxSize work as designed?

One of the most complex part of the app is the layout of the thumbnails. This is handled by `ThumbnailsPanel` that has a `NSCollectionView` member, but implements its `NSCollectionViewDataSource, NSCollectionViewDelegate, NSCollectionViewDelegateFlowLayout` protocols itself. There is also a custom class for `CollectionViewCenterFlowLayout`.

All these moving pieces have implementation hooks you're supposed to use. It is very OOP so you're supposed to subclass or implements protocols in order to customize your `NSCollectionView`. Documentation from Apple as usual is terrible, so I had to jump through many loops to get it exactly the way I wanted.

Indeed, when the panel can become bigger than the screen, instead of stopping at the bottom and providing scroll maybe like you see in the Photos app, or another app from Apple. My point is that the documentation is terrible and I did the current implementation. When we go forward to implementing #69, we can change a lot if there is a better way to approach this.

I'm really struggling to understand how one is supposed to use some parts of Cocoa sometimes

Reply:
@lwouis 

> One of the most complex part of the app is the layout of the thumbnails. 

And thats a part where my lacking Swift/Cocoa knowledge shows itself nicely and i could not explain the flow fully right now. Btw. AppCode does not support Call Hierarchy yet for Swift (bummer!). Had to switch to Xcode to get a grasp at the execution flow with the Call Hierarchy there and where to make the required changes to have the same preferredScreen everywhere.

Regarding the Documentation from Apple: Do you know if one can find the source code of their implementation or is this closed source? E.g. for NSCollectionViewFlowLayout.

I thought this feature would be not that of a big change but it send me all over the codebase (which is good).

> Indeed, when the panel can become bigger than the screen, instead of stopping at the bottom and providing scroll maybe like you see in the Photos app, or another app from Apple. My point is that the documentation is terrible and I did the current implementation. When we go forward to implementing #69, we can change a lot if there is a better way to approach this.
> I'm really struggling to understand how one is supposed to use some parts of Cocoa sometimes. Especially because every tutorial on Earth show how to build UI with InterfaceBuilder, and I'm actively wanting to not use that tool. Maybe it's a mistake though?

For complex / dynamic windows like ThumbnailsPanel i think its right to go the programmatic way but my knowledge here if greatly limited, who knows maybe it is nicely interoperable.

The PreferencesPanel could be probably done easily with InterfaceBuilder (static) but you said there is some dependency on Xcode/Framework versions which would up the min required macOS version for AltTab right?

Reply:
> Do you know if one can find the source code of their implementation or is this closed source

I believe macOS libraries such as Cocoa are all closed-source, but I may be wrong, or some people might have disassembled lots of it by now. Not sure

> there is some dependency on Xcode/Framework versions which would up the min required macOS version for AltTab right?

The reason I avoided InterfaceBuilder is that it's a GUI tool unique to XCode. I want to work in AppCode, and also I want for other people to be apple to contribute even without having XCode installed. The way the project is right now you can use any text editor and edit Swift code. Interface Builder produces either binary or XML files which are blackboxes from a version control point of view, and also when you don't have XCode handy. Like imagine a complex changes to a `.xib` (Interface Builder output) to review in a PR. You now have to checkout the code in XCode in order to have any idea what was done. Then in XCode you need to check all the UI checkboxes, inspectors, etc, and guess what was changed.

No constraints on min macOS version though. That only depends on which OS APIs are called.

Reply:
@lwouis I see now where you are coming from, thanks! Going through every UI objects and its setting panels (there are plenty of them there) manually without VCS would a big no.